alternator/expression_factory.cc

```diff
@@ -0,0 +1,188 @@
+        size_t operator()(const key& k) const noexcept { return std::hash<std::string_view>{}(k.str); }
+        size_t operator()(std::string_view sv) const noexcept { return std::hash<std::string_view>{}(sv); }
+    };
+    struct key_equal {
+        using is_transparent = void;
+        bool operator()(const key& k, std::string_view sv) const noexcept { return k.str == sv; }
+        bool operator()(std::string_view sv, const key& k) const noexcept { return sv == k.str; }
+        bool operator()(const key& a, const key& b) const noexcept { return a.str == b.str; }
+    };
+
+    using string_map = absl::node_hash_map<key, Value, key_hash, key_equal>;
+    string_map map;
+    using lru_key_list_hook = boost::intrusive::member_hook<key, boost::intrusive::list_member_hook<>,&key::list_hook>;
+    using lru_key_list = boost::intrusive::list<key, lru_key_list_hook>;
+    lru_key_list lru_list;
+
+    // Move the accessed key to the back of the LRU list
+    template<bool IsNew=false> 
+    void touch(string_map::iterator it) {
+        // key is const, but it has mutable list_hook, so we remove constness
+        // to allow moving it in the list.
+        // This is safe because we only use this for moving keys in the LRU list,
+        // and we never modify the key itself.
+        key& key_ref = const_cast<key&>(it->first);
+        if constexpr (!IsNew) {
+            lru_list.erase(lru_list.iterator_to(key_ref));
+        }
+        lru_list.push_back(key_ref);
+    }
+public:
+    auto size() { return map.size(); }
+    void reserve(size_t n) { return map.reserve(n); }
+
+    // Lookup by string_view, move key to back, and return value
+    std::optional<std::reference_wrapper<Value>> find(std::string_view key) {
+        auto it = map.find(key);
+        if(it == map.end()) {
+            return std::nullopt;
+        }
+        touch(it);
+        return std::ref(it->second);
+    }
+
+    // Removes the least recently used item from the cache
+    // Returns true if an item was removed, false if the cache was empty
+    bool pop() {
+        if (!lru_list.empty()) {
+            auto it = map.find(lru_list.front());
+            lru_list.pop_front();
+            if(it != map.end()) { // This should be always true
+                map.erase(it);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    // Insert or update the value for the given key.
+    void push(std::string_view key, Value&& value) {
+        auto ret = map.insert_or_assign(key, std::move(value));
+        (ret.second ? touch<true>(ret.first) : touch<false>(ret.first));
+    }
+};
+}
+
+// Actual parsing functions are defined in alternator/expressions.cc
+parsed::update_expression parse_update_expression(std::string_view query);
+std::vector<parsed::path> parse_projection_expression(std::string_view query);
+parsed::condition_expression parse_condition_expression(std::string_view query, const char* caller);
```

Comment:
These declaration should be in a header file used by both this file and alternator/expression.cc - it helps check against differing signatures.
They were actually already in alternator/expression.hh. Did you remove them?